File: day-10/python/main.py
import pathlib
from collections import defaultdict


def find_joltage_differences_product(adapters: list[int]) -> int:
    jolt_differences = defaultdict(int)
    for index, adapter_jolts in enumerate(adapters[:-1]):
        jolt_differences[adapters[index + 1] - adapter_jolts] += 1
    return jolt_differences[1] * jolt_differences[3]


def find_num_valid_permutations(adapters: list[int]) -> int:
    num_valid_permutations = {0: 1}
    for adapter_jolts in adapters[1:]:
        num_valid_permutations[adapter_jolts] = sum(
            num_valid_permutations.get(adapter_jolts - delta, 0)
            for delta in range(1, 4)
        )
    return num_valid_permutations[adapters[-1]]


with open(pathlib.Path(__file__).parent.parent / 'input.txt') as f:
    adapters = sorted(int(line) for line in f.read().splitlines())

adapters.insert(0, 0)
adapters.append(adapters[-1] + 3)

part_one_solution = find_joltage_differences_product(adapters)
part_two_solution = find_num_valid_permutations(adapters)

print('Part One:', part_one_solution)
print('Part Two:', part_two_solution)

assert part_one_solution == 2112
assert part_two_solution == 3022415986688

File: day-10/rust/src/main.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

fn find_jolt_differences_product(adapter_jolts: &[i64]) -> i64 {
    let mut jolt_differences = HashMap::new();
    for (index, jolts) in adapter_jolts[..adapter_jolts.len() - 1].iter().enumerate() {
        *jolt_differences
            .entry(adapter_jolts[index + 1] - jolts)
            .or_insert(0) += 1;
    }
    jolt_differences[&1] * jolt_differences[&3]
}

fn find_valid_permutations(adapter_jolts: &[i64]) -> i64 {
    let mut num_valid_permutations: HashMap<_, _> = [(0, 1)].iter().cloned().collect();
    for jolts in adapter_jolts[1..].iter() {
        num_valid_permutations.insert(
            *jolts,
            (1..4)
                .map(|delta| num_valid_permutations.get(&(jolts - delta)).unwrap_or(&0))
                .sum(),
        );
    }
    num_valid_permutations[&adapter_jolts.last().unwrap()]
}

fn main() {
    let mut adapter_jolts: Vec<_> = fs::read_to_string(
        Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("input.txt"),
    )
    .unwrap()
    .lines()
    .map(|l| l.parse().unwrap())
    .collect();

    adapter_jolts.sort_unstable();
    adapter_jolts.insert(0, 0);
    adapter_jolts.push(adapter_jolts.last().unwrap() + 3);

    let part_one_solution = find_jolt_differences_product(&adapter_jolts);
    let part_two_solution = find_valid_permutations(&adapter_jolts);

    println!("Part One: {}", &part_one_solution);
    println!("Part Two: {}", &part_two_solution);

    assert_eq!(part_one_solution, 2112);
    assert_eq!(part_two_solution, 3022415986688);
}
